src/components/include/sl.h

```diff
@@ -438,22 +439,41 @@ sl_cs_exit_schedule_nospin_arg(struct sl_thd *to)
 
 	if (t->properties & SL_THD_PROPERTY_OWN_TCAP) {
 		assert(t->budget && t->period);
+		assert(sl_thd_tcap(t) != sl__globals()->sched_tcap);
 
 		if (t->last_replenish == 0 || t->last_replenish + t->period <= now) {
-			tcap_res_t currbudget;
+			tcap_res_t currbudget = 0;
+			cycles_t replenish    = now - ((now - t->last_replenish) % t->period);  
 
-			t->last_replenish = now;
-			currbudget        = (tcap_res_t)cos_introspect(ci, sl_thd_tcap(t), TCAP_GET_BUDGET);
-			/* TODO: need to change logic for SNDCAP with tcap_delegate, and error handling */
-			if (currbudget < t->budget && cos_tcap_transfer(sl_thd_rcvcap(t), sl__globals()->sched_tcap, (t->budget - currbudget), t->prio)) assert(0);
+			ret = 0;
+			if (likely(t->last_replenish)) currbudget = (tcap_res_t)cos_introspect(ci, sl_thd_tcap(t), TCAP_GET_BUDGET);
+
+			if (!cycles_same(currbudget, t->budget, SL_CYCS_DIFF) && currbudget < t->budget) {
+				tcap_res_t transfer = t->budget - currbudget;
+
+				ret = cos_tcap_transfer(sl_thd_rcvcap(t), sl__globals()->sched_tcap, transfer, t->prio);
+			}
+
+			if (likely(ret == 0)) t->last_replenish = replenish;
```

Comment:
This is all a lot of logic and conditionals.  Things are getting complicated, and more expensive.

It is all broken as well because we rely on introspection where the child could have transferred the budget away in the mean time.  It seems like this all really needs a discussion after the deadline.

Reply:
Yes, I agree.